nav2_map_server/include/nav2_map_server/vector_object_server.hpp

```diff
@@ -0,0 +1,229 @@
+   */
+  void getMapBoundaries(
+    double & min_x, double & min_y, double & max_x, double & max_y) const;
+
+  /**
+   * @brief Creates or updates existing map with required sizes and fills it with default value
+   * @param min_x min X-boundary of new map
+   * @param min_y min Y-boundary of new map
+   * @param max_x max X-boundary of new map
+   * @param max_y max Y-boundary of new map
+   * @throw std::exception if map has negative X or Y size
+   */
+  void updateMap(
+    const double & min_x, const double & min_y, const double & max_x, const double & max_y);
+
+  /**
+   * @brief Processes all vector objects on raster output map
+   */
+  void putVectorObjectsOnMap();
+
+  /**
+   * @brief Publishes output map
+   */
+  void publishMap();
+
+  /**
+   * @brief Calculates new map sizes, updates map, processes all vector objects on it
+   * and publishes output map one time
+   */
+  void processMap();
+
+  /**
+   * @brief If map to be update dynamically, creates map processing timer,
+   * otherwise process map once
+   */
+  void switchMapUpdate();
+
+  /**
+   * @brief Callback for AddShapes service call.
+   * Reads all input vector objects from service request,
+   * creates them or updates their shape in case of existing objects
+   * and switches map processing/publishing routine
+   * @param request_header Service request header
+   * @param request Service request
+   * @param response Service response
+   */
+  void addShapesCallback(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<nav2_msgs::srv::AddShapes::Request> request,
+    std::shared_ptr<nav2_msgs::srv::AddShapes::Response> response);
+
+  /**
+   * @brief Callback for GetShapes service call.
+   * Gets all shapes and returns them to the service response
+   * @param request_header Service request header
+   * @param request Service request
+   * @param response Service response
+   */
+  void getShapesCallback(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<nav2_msgs::srv::GetShapes::Request> request,
+    std::shared_ptr<nav2_msgs::srv::GetShapes::Response> response);
+
+  /**
+   * @brief Callback for RemoveShapes service call.
+   * Try to remove requested vector objects and switches map processing/publishing routine
+   * @param request_header Service request header
+   * @param request Service request
+   * @param response Service response
+   */
+  void removeShapesCallback(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<nav2_msgs::srv::RemoveShapes::Request> request,
+    std::shared_ptr<nav2_msgs::srv::RemoveShapes::Response> response);
+
+protected:
+  /// @brief TF buffer
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+  /// @brief TF listener
+  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
+
+  /// @brief All shapes vector
+  std::vector<std::shared_ptr<Shape>> shapes_;
```

Comment:
I don't think these need to be shared pointers after looking over the code, no? They could be unique pointers